### PR TITLE
Implement `scan database` command

### LINKED ISSUE #161: Add a `database-coverage` command
After thinking about some ways to reduce cognitive load on users, I think we need an _optional_ command that would allow users to verify that all of the fields in their databases are included in their dataset manifest files. This would then throw an automatic error when a user adds a dataset/collection/field but doesn't document it in their manifests

There are two possible options that I can think of for implementation here (please suggest other options as you think of them!):
1. user can pass a map of `{fides_key: database_url}` via the config/env vars
2. we could store these credentials (safely) within the API/db

I lean towards #1 from a design perspective, not to mention if we can avoid storing sensitive credentials in the db we should at all costs

**Example 1 for embedding the validation values into the TOML file**
```toml
[cli]
dataset_validation = {dataset_1 = "<database_url_1>", dataset_2 = "<database_url_2>"}
```

**Example 2 for embedding the validation values into the TOML file**
```toml
[datasets]
dataset_1 = "database_url_1"
dataset_2 = "database_url_2"
```

### PR DESCRIPTION
**Pull request by earmenda**
Closes #161 

### Code Changes

* [x] Implement `find_missing_dataset_fields` function that iterates through db collections and finds missing dataset fields
* [x] Implement `database_coverage` function that fetches a dataset and db collections and outputs results of `find_missing_dataset_fields`

### Steps to Confirm

* [x] Unit tests test `find_missing_dataset_fields` extensively
* [x] Could use a simple integration test but it might be better to wait for #289 that writes new tests for generate-dataset

### Pre-Merge Checklist

* [x] All CI Pipelines Succeeded
* [x] Documentation Updated

### Description Of Changes

* Follows examples from #161

### CODE REVIEW

**Comment by earmenda** [comment]
@ThomasLaPiana @NevilleS @iamkelllly - here is my first pass on the implementation of the `database-coverage` command. Can you go through these command outputs and let me know if this is what you had in mind? 

### Command Usage

Created a table with this schema:
```
CREATE SCHEMA coverage
  CREATE TABLE mytable (
    name        varchar(40) CONSTRAINT firstkey PRIMARY KEY,
    phone       varchar(40),
    address     varchar(40)
  );
```
Run generate-dataset command:
```
fidesctl generate-dataset "postgresql+psycopg2://postgres:fidesctl@fidesctl-db:5432/fidesctl_test" test2.yml
```

That creates a dataset like this:
```
dataset:
- fides_key: coverage
  organization_fides_key: default_organization
  name: coverage
  description: 'Fides Generated Description for Schema: coverage'
  meta: null
  data_categories: null
  data_qualifier: aggregated.anonymized.unlinked_pseudonymized.pseudonymized.identified
  collections:
  - name: coverage.mytable
    description: 'Fides Generated Description for Table: coverage.mytable'
    data_categories: null
    data_qualifier: aggregated.anonymized.unlinked_pseudonymized.pseudonymized.identified
    fields:
    - name: address
      description: 'Fides Generated Description for Column: address'
      data_categories: []
      data_qualifier: aggregated.anonymized.unlinked_pseudonymized.pseudonymized.identified
    - name: name
      description: 'Fides Generated Description for Column: name'
      data_categories: []
      data_qualifier: aggregated.anonymized.unlinked_pseudonymized.pseudonymized.identified
    - name: phone
      description: 'Fides Generated Description for Column: phone'
      data_categories: []
      data_qualifier: aggregated.anonymized.unlinked_pseudonymized.pseudonymized.identified
```

Then apply the dataset:
```
root@338ffc45e5d9:/fides/fidesctl# fidesctl apply test2.yml 
Loading resource manifests from: test2.yml
Taxonomy successfully created.
----------
Processing dataset resources...
CREATED 1 dataset resources.
UPDATED 0 dataset resources.
SKIPPED 0 dataset resources.
----------
```

We can check `database-coverage` now which should have 100% coverage for data set `coverage`
```
root@338ffc45e5d9:/fides/fidesctl# fidesctl database-coverage "postgresql+psycopg2://postgres:fidesctl@fidesctl-db:5432/fidesctl_test" coverage
`coverage` annotation coverage: 100% 
```

Now we add a column:
```
ALTER TABLE coverage.mytable
ADD COLUMN test_adding_column varchar(40),
ADD COLUMN test_adding_column2 varchar(40);
```

Checking `database-coverage` now will report the two missing columns. This also is printed in red and exits in error:
```
root@338ffc45e5d9:/fides/fidesctl# fidesctl database-coverage "postgresql+psycopg2://postgres:fidesctl@fidesctl-db:5432/fidesctl_test" coverage
`coverage` annotation coverage: 60% 
The following fields do not have any data category annotations: 
coverage.coverage.mytable.test_adding_column
coverage.coverage.mytable.test_adding_column2
```

We can run the same command but with a `coverage-threshold` option that allows 50% coverage. The output is the same but the output prints in green and does not exit in error. 
```
root@338ffc45e5d9:/fides/fidesctl# fidesctl  database-coverage --coverage-threshold .5 "postgresql+psycopg2://postgres:fidesctl@fidesctl-db:5432/fidesctl_test" coverage
`coverage` annotation coverage: 60% 
The following fields do not have any data category annotations: 
coverage.coverage.mytable.test_adding_column
coverage.coverage.mytable.test_adding_column2
```

### Error Modes

Invalid coverage-threshold:
```
root@338ffc45e5d9:/fides/fidesctl# fidesctl  database-coverage --coverage-threshold 2 "postgresql+psycopg2://postgres:fidesctl@fidesctl-db:5432/fidesctl_test" coverage
Usage: fidesctl database-coverage [OPTIONS] CONNECTION_STRING DATASET_KEY
Try 'fidesctl database-coverage -h' for help.

Error: Invalid value for '-c' / '--coverage-threshold': 2.0 is not in the valid range of 0 to 1.
```

Dataset key provided does not exist in fidesctl database:
```
root@338ffc45e5d9:/fides/fidesctl# fidesctl database-coverage "postgresql+psycopg2://postgres:fidesctl@fidesctl-db:5432/fidesctl_test" coverage
Dataset (coverage) does not exist in existing taxonomy
```

Dataset exists in fidesctl database but there is no matching dataset in remote database:
```
root@338ffc45e5d9:/fides/fidesctl# fidesctl database-coverage "postgresql+psycopg2://postgres:fidesctl@fidesctl-db:5432/fidesctl_test" coverage_missing
Dataset (coverage_missing) could not be found in database. Found datasets (dict_keys(['coverage', 'public']))
```

### Open Questions? 
* Should we take the time now to add more documentation around dataset commands? Currently the dataset.md tutorial page only has a small section for `generate-dataset` and has this note: `<!-- TODO: Link to the `generate-dataset` usage documentation below, when it exists. -->`. We could add another short note regarding `database-coverage` or write alot more documentation around it now. 
* Thoughts on us only fetching the given dataset from the fidesctl server? This makes the most sense to me but we could work with yaml files as well if we wanted.

**Comment by NevilleS** [comment]
@earmenda question here: are you counting coverage as "field is defined" or are you counting coverage as "field is defined & has at least one data category annotation"? It looks like the former to me based on the example of generating a dataset then evaluating coverage is immediately 100%.

My preference is that a dataset with zero data categories is treated as 0% coverage, since the job we're trying to achieve with dataset annotations is to understand the data categories within them. So as implemented this is close but I think we should make it more strict

**Comment by ThomasLaPiana** [comment]
As per a slack discussion, the requirements are now this:

dataset and collections-level data categories don't count towards coverage. The reason being that it would be very easy to add new columns and never actually have coverage change. We want to be explicit that each column has a data category. The only two criteria for coverage are:

1. The field is defined
2. The field has at least one data category

cc @earmenda

**Comment by earmenda** [comment]
Made changes around checking for categories. Categories are not inherited from Dataset or Collections when it comes to coverage. 

Also merged in main and fixed some merge issues.

This is mergeable as is but I do have some things on my mind for this still:
* `call it fidesctl database-scan instead? (I also like the fidesctl scan database … suggestion)` - From Neville
* add a simple integration test per db?
* noted documentation is lacking right now for dataset, should we worry about that for now?

**Comment by ThomasLaPiana** [comment]
> Made changes around checking for categories. Categories are not inherited from Dataset or Collections when it comes to coverage.
> 
> Also merged in main and fixed some merge issues.
> 
> This is mergeable as is but I do have some things on my mind for this still:
> 
> * `call it fidesctl database-scan instead? (I also like the fidesctl scan database … suggestion)` - From Neville
> * add a simple integration test per db?
> * noted documentation is lacking right now for dataset, should we worry about that for now?

I think getting those 3 things in would put the cherry on top of this PR:
1. Let's do the `fidesctl scan database` route. That will require a new subcommand `scan` but shouldn't be too difficult. Main thing is to make sure it doesn't break the auto-generated documentation
2. Yep, might as well since the plumbing is already there for that i think?
3. Yeah, I'm happy to jump in here as well and write some of these docs

**Comment by earmenda** [comment]
Updated command to be called in this format:
`scan database "postgresql+psycopg2://postgres:fidesctl@fidesctl-db:5432/fidesctl_test" coverage`

The caveat here is that I had some difficulties creating a `scan` group with child commands. For the moment, database is an argument of the scan command. I didn't want to hold this up and functionally it is about the same. Let me know if this is good enough for now and we can add a new `scan` group when we need additional scan targets.

**Comment by earmenda** [comment]
Sample usages after latest changes:

```
fidesctl scan --manifest-dir fidesctl/fides_resources database "postgresql+psycopg2://postgres:fidesctl@fidesctl-db:5432/fidesctl_test"
```

```
fidesctl scan database "postgresql+psycopg2://postgres:fidesctl@fidesctl-db:5432/fidesctl_test"
```

```
fidesctl scan --coverage-threshold 5  --manifest-dir fidesctl/fides_resources database "postgresql+psycopg2://postgres:fidesctl@fidesctl-db:5432/fidesctl_test"
```

* Dataset key no longer an input. All datasets are checked and output is aggregated for all
* Added a manifest-dir which allows you to pass in a file instead of pulling datasets from the server
* coverage threshold is now an interget 0-100

**Comment by earmenda** [comment]
I didnt really get around to finishing documentation. Could potentially also add an integration test for new manifest-dir option

**Comment by ThomasLaPiana** [comment]
ran this and it worked as expected, throwing errors and when updating the dataset, _not_ throwing errors as expected

`fidesctl scan -m fides_resources database postgresql+psycopg2://postgres:fidesctl@fidesctl-db:5432/fidesctl`

**Comment by SteveDMurphy** [comment]
Everything is working for me here as well from what I can tell, is it just the docs and integration test that are left on this then?

File: fidesctl/src/fidesctl/core/generate_dataset.py
Comment (PSalant726) [inline fidesctl/src/fidesctl/core/generate_dataset.py:264]:
```suggestion
    output = "Successfully scanned the following datasets:\n"
    output += f"\t{'\n\t'.join(list(db_collections.keys()).sort())}\n"
    print(output)
    
    if uncategorized_fields:
        uncategorized_output = "The following fields are missing data category annotations:\n"
        uncategorized_output += f"\t{'\n\t'.join(uncategorized_fields.sort())}\n"
        print(uncategorized_output)

    annotation_output = f"Annotation coverage: {coverage_percent}%\n\n"
    if coverage_percent < coverage_threshold:
        echo_red(annotation_output)
        raise SystemExit(1)
    echo_green(annotation_output)
```

This might need some tweaking. I'm going for output like:

```
Successfully scanned the following datasets:
    dataset_key_one
    dataset_key_three
    dataset_key_two

The following fields are missing data category annotations:
    uncategorized_field_one
    uncategorized_field_three
    uncategorized_field_two

Annotation coverage: 85%   # Either red or green, exits with a 1 status when red
```

Separately, and regardless of whether you feel this is an improved output format, it would be nice to extract the output construction into a helper function such that this consolidates to:

```python
print(
    results(db_collection.keys(), uncategorized_fields, coverage_percent, coverage_threshold)
)
```

File: fidesctl/src/fidesctl/core/generate_dataset.py
Comment (PSalant726) [inline fidesctl/src/fidesctl/core/generate_dataset.py:186]:
```suggestion
    Finds all uncategorized fields given a database modeled object. Datasets
```

**Review by PSalant726 — CHANGES_REQUESTED** [review]

File: fidesctl/src/fidesctl/cli/cli.py
Comment (SteveDMurphy) [reply fidesctl/src/fidesctl/cli/cli.py:146]: I'd like to see something different than a raw connection string as well, at least eventually (not sure around the timing for getting this functionality out), and I really like the idea of decomposing into components 👍🏽  I'm still not confident in having a wider understanding as yet, but if we can imagine having 3 to 5 connection strings to maintain I would prefer to have a config.yml/json that allows me to manage those as strings or env vars as needed and reference the connection by name/key instead

**Review by SteveDMurphy — COMMENTED** [review]
Overall great work on everything @earmenda, especially with all of the conversations around it right up to the holiday! 🙌🏽  

I got the benefit of reading the details from @PSalant726 which were super helpful for me, the only thing I was going to add was around the issues we ran into with the grouped click command. Functionally it is the same as you mentioned, but I ended up having to pass the options in a different position than if we were able to resolve the grouped click command here versus down the road. I think there is a path to getting it working but wanted to get thoughts on whether or not it is important to resolve today or if down the road is still fine at this point.

So excited to see this coming through as it was one of the things I asked for in my beta exit chat!!

File: fidesctl/src/fidesctl/core/generate_dataset.py
Comment (earmenda) [reply fidesctl/src/fidesctl/core/generate_dataset.py:92]: Yeah that seems like a good idea. It's a little verbose because the type is `Dict[str, Callable[[Engine], Dict[str, Dict[str, List[str]]]]]`

Verified that if one of the functions returns something different then we have a mypy failure:
```
root@d12908839f6b:/fides/fidesctl# mypy
src/fidesctl/core/generate_dataset.py:93: error: Dict entry 0 has incompatible type "str": "Callable[[Engine], Dict[str, Dict[str, List[int]]]]"; expected "str":
"Callable[[Engine], Dict[str, Dict[str, List[str]]]]"  [dict-item]
            "postgresql": get_postgres_collections_and_fields,
            ^
Found 1 error in 1 file (checked 36 source files)
```

File: fidesctl/src/fidesctl/core/generate_dataset.py
Comment (earmenda) [reply fidesctl/src/fidesctl/core/generate_dataset.py:153]: fixed

File: fidesctl/src/fidesctl/core/generate_dataset.py
Comment (PSalant726) [reply fidesctl/src/fidesctl/core/generate_dataset.py:92]: Personally I think it's worth the ugliness. Thanks for investigating!

File: fidesctl/src/fidesctl/core/generate_dataset.py
Comment (earmenda) [reply fidesctl/src/fidesctl/core/generate_dataset.py:175]: Ah, `field_found` was a refactoring miss because we were originally just checking for that. Thanks.

I thought about your suggestion and was on the fence at first but i do think it's most logical to be searching for `field_uncategorized` as opposed to `field_categorized`.  Only thing missing is the change to `else True`

File: fidesctl/src/fidesctl/cli/cli.py
Comment (ThomasLaPiana) [reply fidesctl/src/fidesctl/cli/cli.py:146]: in full agreement here that this isn't optimal, but this would be another issue in and of itself and we use other functions that also rely on a user-defined complete connection string

will also definitely need to do some kind of mapping like @SteveDMurphy mentioned (i gave some examples in the issue) but those are outside of the scope of this issue, this is just an MVP and I think its ok to use the full connection string in that context until we expand the feature-set further

File: fidesctl/src/fidesctl/cli/cli.py
Comment (ThomasLaPiana) [reply fidesctl/src/fidesctl/cli/cli.py:169]: sanitization sounds like a good idea, but what is the actual attack vector here? Do we need to sanitize every input for a CLI that only runs locally?

**Comment by ThomasLaPiana** [comment]
> Overall great work on everything @earmenda, especially with all of the conversations around it right up to the holiday! 🙌🏽
> 
> I got the benefit of reading the details from @PSalant726 which were super helpful for me, the only thing I was going to add was around the issues we ran into with the grouped click command. Functionally it is the same as you mentioned, but I ended up having to pass the options in a different position than if we were able to resolve the grouped click command here versus down the road. I think there is a path to getting it working but wanted to get thoughts on whether or not it is important to resolve today or if down the road is still fine at this point.
> 
> So excited to see this coming through as it was one of the things I asked for in my beta exit chat!!

I think finessing the click commands can be done in a smaller issue, this is already a huge change and the potential risk of reworking the click commands feels too high to me

File: fidesctl/src/fidesctl/cli/cli.py
Comment (PSalant726) [reply fidesctl/src/fidesctl/cli/cli.py:169]: Yes, we should always validate all user input. It's my understanding that the CLI runs locally, but may also be used in CI environments. Regardless, acceptance of arbitrary data is an attack vector even if there's no immediately obvious vulnerability. The code to handle this input will change over time, but the requirements will not change - these particular strings would still need to be a valid database connection string and an existing directory path.

We can short-circuit the command by throwing if `os.path.isdir(manifest_dir) == False` and/or if `connection_string` is not a valid URL, although a cursory search for URL validators outside of Django environments isn't showing much other than [this SO answer](https://stackoverflow.com/questions/7160737/how-to-validate-a-url-in-python-malformed-or-not#answer-7160778), which would require some editing.

Based on https://github.com/ethyca/fides/pull/292#discussion_r778314126, it sounds like we might want to create a separate issue to handle validation of all user input throughout the CLI.

File: fidesctl/src/fidesctl/core/generate_dataset.py
Comment (ThomasLaPiana) [reply fidesctl/src/fidesctl/core/generate_dataset.py:252]: yeah if someone is explicitly checking a database and it comes back empty, that seems like a problem and we'd want to surface that

File: fidesctl/src/fidesctl/core/generate_dataset.py
Comment (earmenda) [reply fidesctl/src/fidesctl/core/generate_dataset.py:186]: fixed

**Comment by ThomasLaPiana** [comment]
@earmenda final things to change before merge!

1. https://github.com/ethyca/fides/pull/292/files#r777722308
2. https://github.com/ethyca/fides/pull/292#discussion_r777733725

Once we surface empty databases and slightly update the output format/info, i think this looks great

File: fidesctl/src/fidesctl/core/generate_dataset.py
Comment (earmenda) [reply fidesctl/src/fidesctl/core/generate_dataset.py:175]: woops, should also be `or` so... `field.name != db_dataset_field or not field.data_categories`

File: fidesctl/src/fidesctl/core/generate_dataset.py
Comment (earmenda) [reply fidesctl/src/fidesctl/core/generate_dataset.py:252]: I added this within the same method for now. Adding it to a method like `find_all_uncategorized_dataset_fields` feels bad since these methods could be used for other purposes.

File: fidesctl/src/fidesctl/core/generate_dataset.py
Comment (earmenda) [reply fidesctl/src/fidesctl/core/generate_dataset.py:264]: good change, updated

**Comment by earmenda** [comment]
> Overall great work on everything @earmenda, especially with all of the conversations around it right up to the holiday! 🙌🏽
> 
> I got the benefit of reading the details from @PSalant726 which were super helpful for me, the only thing I was going to add was around the issues we ran into with the grouped click command. Functionally it is the same as you mentioned, but I ended up having to pass the options in a different position than if we were able to resolve the grouped click command here versus down the road. I think there is a path to getting it working but wanted to get thoughts on whether or not it is important to resolve today or if down the road is still fine at this point.
> 
> So excited to see this coming through as it was one of the things I asked for in my beta exit chat!!

sounds good to me. As we add more `scan` commands this should be a priority because it will make the code alot cleaner. I don't want to have a `scan` method that is really overloaded.

**Comment by earmenda** [comment]
Thanks for all the comments @PSalant726, i appreciate the detailed feedback

**Comment by earmenda** [comment]
> Overall great work on everything @earmenda, especially with all of the conversations around it right up to the holiday! 🙌🏽
> 
> I got the benefit of reading the details from @PSalant726 which were super helpful for me, the only thing I was going to add was around the issues we ran into with the grouped click command. Functionally it is the same as you mentioned, but I ended up having to pass the options in a different position than if we were able to resolve the grouped click command here versus down the road. I think there is a path to getting it working but wanted to get thoughts on whether or not it is important to resolve today or if down the road is still fine at this point.
> 
> So excited to see this coming through as it was one of the things I asked for in my beta exit chat!!

thanks :)

**Comment by earmenda** [comment]
@ThomasLaPiana 

Last thing left is just the updated docs, will get that out tonight so we could release tomorrow.

**Review by PSalant726 — APPROVED** [review]

**Comment by earmenda** [comment]
Sample output after the latest changes:

```
root@16fadc5e1da4:/fides/fidesctl# fidesctl scan database "postgresql+psycopg2://postgres:fidesctl@fidesctl-db:5432/fidesctl_test"
Successfully scanned the following datasets:
        public

The following fields are missing data category annotations:
        public.public.alembic_version.version_num
        public.public.data_categories.id
        public.public.data_categories.fides_key
        public.public.data_categories.organization_fides_key
        public.public.data_categories.name
        public.public.data_categories.description
        public.public.data_categories.parent_key
        public.public.data_categories.created_at
        public.public.data_categories.updated_at
        public.public.data_qualifiers.id
        public.public.data_qualifiers.fides_key
        public.public.data_qualifiers.organization_fides_key
        public.public.data_qualifiers.name
        public.public.data_qualifiers.description
        public.public.data_qualifiers.parent_key
        public.public.data_qualifiers.created_at
        public.public.data_qualifiers.updated_at
        public.public.data_subjects.id
        public.public.data_subjects.fides_key
        public.public.data_subjects.organization_fides_key
        public.public.data_subjects.name
        public.public.data_subjects.description
        public.public.data_subjects.created_at
        public.public.data_subjects.updated_at
        public.public.data_uses.id
        public.public.data_uses.fides_key
        public.public.data_uses.organization_fides_key
        public.public.data_uses.name
        public.public.data_uses.description
        public.public.data_uses.parent_key
        public.public.data_uses.created_at
        public.public.data_uses.updated_at
        public.public.datasets.id
        public.public.datasets.fides_key
        public.public.datasets.organization_fides_key
        public.public.datasets.name
        public.public.datasets.description
        public.public.datasets.meta
        public.public.datasets.data_categories
        public.public.datasets.collections
        public.public.datasets.data_qualifier
        public.public.datasets.created_at
        public.public.datasets.updated_at
        public.public.organizations.id
        public.public.organizations.fides_key
        public.public.organizations.organization_fides_key
        public.public.organizations.name
        public.public.organizations.description
        public.public.organizations.organization_parent_key
        public.public.organizations.created_at
        public.public.organizations.updated_at
        public.public.policies.id
        public.public.policies.fides_key
        public.public.policies.organization_fides_key
        public.public.policies.name
        public.public.policies.description
        public.public.policies.rules
        public.public.policies.created_at
        public.public.policies.updated_at
        public.public.registries.id
        public.public.registries.fides_key
        public.public.registries.organization_fides_key
        public.public.registries.name
        public.public.registries.description
        public.public.registries.created_at
        public.public.registries.updated_at
        public.public.systems.id
        public.public.systems.fides_key
        public.public.systems.organization_fides_key
        public.public.systems.name
        public.public.systems.description
        public.public.systems.registry_id
        public.public.systems.meta
        public.public.systems.system_type
        public.public.systems.system_dependencies
        public.public.systems.privacy_declarations
        public.public.systems.created_at
        public.public.systems.updated_at
        public.public.evaluations.id
        public.public.evaluations.fides_key
        public.public.evaluations.status
        public.public.evaluations.message
        public.public.evaluations.violations
        public.public.evaluations.created_at
        public.public.evaluations.updated_at

Annotation coverage: 0% <---- red
```

**Review by ThomasLaPiana — APPROVED** [review]
Looks great to me :)

**Pull request merged**
